pinc/upload_file.inc

```diff
@@ -0,0 +1,255 @@
+{
+    $file_path = "";
+    try
+    {
+        $mode = get_enumerated_param($_REQUEST, 'mode', 'upload', ['upload', 'resumable']);
+        if('upload' == $mode)
+        {
+            $file_info = $_FILES['uploaded_file'];
+
+            // If a user hits the "Upload" button without first selecting a file,
+            // it appears that most browsers send a request containing a file whose
+            // name and content are empty. But I think it's also legal for a browser
+            // to send a request that doesn't contain a file at all (in which case
+            // $file_info would be null.  Check both possibilities.
+            if (is_null($file_info) || $file_info['name'] == '') {
+                return NULL;
+            }
+            $original_name = $file_info['name'];
+            $file_path = $file_info['tmp_name'];
+            $error = $file_info['error'];
+            $file_size = $file_info['size'];
+
+            if ($error != UPLOAD_ERR_OK) {
+                throw new FileException(get_upload_err_msg($error));
+            }
+            // there must be only one file.
+            $file_count = count($original_name);
+            if ($file_count > 1) {
+                throw new FileException( _("You may only upload one file") );
+            }
+            // allow anti_virus to read it
+            chmod($file_path, 0644);
+        }
+        else
+        {
+            // resumable upload
+            $root_staging_dir = "/tmp/resumable_uploads";
+            $original_name = array_get($_POST, "resumable_filename", "");
+            $identifier = array_get($_POST, "resumable_identifier", "");
+            $hashed_filename = md5($identifier);
+            $file_path = "$root_staging_dir/$hashed_filename";
+            $file_size = filesize($file_path);
+            $file_info = ["name" => $original_name, "tmp_name" => $file_path];
+        }
+        // do some checks.
+        empty_check($file_size);
+        virus_check($file_path);
+        return $file_info;
+    }
+    catch(FileException $e)
+    {
+        if(is_file($file_path))
+        {
+            unlink($file_path);
+        }
+        throw($e);
+    }
+}
+
+function empty_check($file_size)
+{
+    if (0 == $file_size) {
+        throw new FileException(_("File is empty"));
+    }
+}
+
+function zip_check($original_name, $file_path)
+{
+    try
+    {
+        if (substr($original_name, -4) != ".zip")
+        {
+            throw new FileException( _("File name extension must be '.zip'") );
+        }
+        // ensure that it's a valid zip
+        // The extension was already checked and the file is not properly named (it has some temporary name), so we should
+        // disable the extension check.
+        if (!is_valid_zip_file($file_path, true))
+        {
+            throw new FileException(sprintf(_("%s is not a valid zip file"),  $original_name));
+        }
+    }
+    catch(FileException $e)
+    {
+        unlink($file_path);
```

Comment:
It feels weird to have this function remove the thing that was uploaded. I would expect the calling function to do that